[registry] Addons styles loader (#6630)
plone/volto: packages/registry/docs/conceptual-guides/add-on-styles-loader.md

---
myst:
  html_meta:
    "description": "An explanation of the add-ons styles loader in @plone/registry"
    "property=og:description": "An explanation of the add-ons styles loader in @plone/registry"
    "property=og:title": "Add-ons styles loader"
    "keywords": "@plone/registry, registry, add-ons, loader"
---

# Add-ons styles loader

The `@plone/registry` compatible add-ons are able to define styles that should be loaded by the app.

> This is achieved by creating a folder {file}`styles` with a {file}`_main.css` file that serves as entry point.

Reviewer: Why the underscore? It makes it look unimportant when actually it's especially important.

Author: I agree. Let's remove the `_`. `main.css` makes sense too?

Reviewer 2: `_` sorts earlier in a directory listing, but that's not important to naming things well. I often see `_` in front of SCSS files, denoting it as a [partial](https://sass-lang.com/guide/#topic-4). There ain't nothing partial about what's in `_main.css`. I also don't want to hold the shift key to get an underscore when typing docs to mention this file. I'd go with `main.css`.

---
myst:
  html_meta:
    "description": "An explanation of the add-ons styles loader in @plone/registry"
    "property=og:description": "An explanation of the add-ons styles loader in @plone/registry"
    "property=og:title": "Add-ons styles loader"
    "keywords": "@plone/registry, registry, add-ons, loader"
---

# Add-ons styles loader

> The `@plone/registry` compatible add-ons are able to declare styles that should be loaded by the app.

Reviewer 2: ```suggestion
Add-ons that are compatible with the `@plone/registry` may declare styles that should be loaded by the app.
```

> This is achieved by creating a folder {file}`styles` with a {file}`main.css` file that serves as entry point.

Reviewer 2: Please verify where the file should be placed. I guessed. Also we can merge steps of creating a directory and file into just a file. I can do that via command line or in PyCharm's file browser.

```suggestion
To do so, create a file {file}`styles/main.css` at the root of your project which serves as the entry point.
```

> This file is a `.css` file containing the styles that we want our app to load.

Reviewer 2: ```suggestion
This file is a `.css` file containing the styles that you want your app to load.
```

> `@plone/registry` has a helper utility `createAddonsStyleLoader` for generating an add-ons style loader file that contains the aggregated files from all the registered add-ons, keeping the order in which they were defined.

Reviewer 2: Split run-on into two sentences.

```suggestion
`@plone/registry` has a helper utility `createAddonsStyleLoader` which generates an add-ons loader file.
That file contains the aggregated files from all the registered add-ons, keeping the order in which they were defined.
```

This loader is also a `.css` file and is placed in the root of your application.
By default, it's called {file}`.addons.styles.css`.

```{important}
This file is generated and maintained by `@plone/registry`.

> You should not modify it and add your own styles in here.

Reviewer 2: ```suggestion
You should neither modify it nor add your own styles in here.
```

> It will be overwriten in the next bundler run.

Reviewer 2: ```suggestion
It will be overwritten in the next bundler run.
```

```

The add-ons loader generator is meant to be run before bundling your app or by the bundler when it runs.

> The `@plone/registry` Vite plugin generates this file so the framework can load it on app bootstrap time.

Reviewer 2: ```suggestion
The `@plone/registry` Vite plugin generates this file, so the framework can load it during app bootstrap time.
```

```js
  const projectRootPath = path.resolve('.');
  const { registry, shadowAliases } = AddonRegistry.init(projectRootPath);

  createAddonsStyleLoader(registry.getAddonStyles());
```

This will create {file}`.addons.styles.css` in the root of your app.

> Afterwards, you have to make sure that your app loads the css, in the preferred way of the framework to do it properly, centralized, and in a performant way.

Reviewer 2: ```suggestion
Afterwards, configure your app to load the CSS according to the framework's convention, and in both a centralized and performant manner.
```

```css
@import "tailwind";
@import "./.addons.styles.css"
```

```{note}

> If you are using a Vite-powered framework, just use the `@plone/registry` Vite plugin.

Reviewer 2: ```suggestion
If you use a Vite-powered framework, you can use the `@plone/registry` Vite plugin as a convenience.
```

Author: same concern as above.

Reviewer 2: Same.

> Only in the case that you are using a non-Vite framework, you will have to build your own integration.

Reviewer 2: ```suggestion
If you use a non-Vite framework, you will have to build your own integration.
```
